lightning/src/routing/router.rs

```diff
@@ -223,77 +383,193 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
-						} else {
-							// max_value means we'll always fail the old_entry.0 > total_fee check
-							total_fee = u64::max_value();
-						}
+			if has_sufficient_liquidity && can_cover_following_hops && amount_to_transfer_over_msat >= $directional_info.htlc_minimum_msat {
+				let hm_entry = dist.entry(&$src_node_id);
+				let old_entry = hm_entry.or_insert_with(|| {
+					// If there was previously no known way to access
+					// the source node (recall it goes payee-to-payer) of $chan_id, first add
+					// a semi-dummy record just to compute the fees to reach the source node.
+					// This will affect our decision on selecting $chan_id
+					// as a way to reach the $dest_node_id.
+					let mut fee_base_msat = u32::max_value();
+					let mut fee_proportional_millionths = u32::max_value();
+					if let Some(Some(fees)) = network.get_nodes().get(&$src_node_id).map(|node| node.lowest_inbound_channel_fees) {
+						fee_base_msat = fees.base_msat;
+						fee_proportional_millionths = fees.proportional_millionths;
 					}
-					let new_graph_node = RouteGraphNode {
-						pubkey: $src_node_id,
-						lowest_fee_to_peer_through_node: total_fee,
-						lowest_fee_to_node: $starting_fee_msat as u64 + new_fee,
-					};
-					if old_entry.0 > total_fee {
-						targets.push(new_graph_node);
-						old_entry.0 = total_fee;
-						old_entry.3 = RouteHop {
+					PathBuildingHop {
+						route_hop: RouteHop {
 							pubkey: $dest_node_id.clone(),
 							node_features: NodeFeatures::empty(),
-							short_channel_id: $chan_id.clone(),
+							short_channel_id: 0,
 							channel_features: $chan_features.clone(),
-							fee_msat: new_fee, // This field is ignored on the last-hop anyway
-							cltv_expiry_delta: $directional_info.cltv_expiry_delta as u32,
+							fee_msat: 0,
+							cltv_expiry_delta: 0,
+						},
+						available_liquidity_msat: 0,
+						src_lowest_inbound_fees: RoutingFees {
+							base_msat: fee_base_msat,
+							proportional_millionths: fee_proportional_millionths,
+						},
+						channel_fees: $directional_info.fees,
+						next_hops_fee_msat: u64::max_value(),
+						hop_use_fee_msat: u64::max_value(),
+						prev_hop_use_estimate_fee_msat: u64::max_value(),
+					}
+				});
+
+				let mut hop_use_fee_msat = 0;
+				let mut prev_hop_use_estimate_fee_msat = 0;
+				let mut total_fee_msat = $next_hops_fee_msat;
+
+				// Ignore hop_use_fee_msat for channel-from-us as we assume all channels-from-us
+				// will have the same effective-fee
+				if $src_node_id != *our_node_id {
+					match compute_fees(amount_to_transfer_over_msat, $directional_info.fees) {
+						// max_value means we'll always fail
+						// the old_entry.get_fee_weight_msat() > total_fee_msat check
+						None => total_fee_msat = u64::max_value(),
+						Some(fee_msat) => {
+							hop_use_fee_msat = fee_msat;
+							total_fee_msat += hop_use_fee_msat;
+							if let Some(prev_hop_fee_msat) = compute_fees(total_fee_msat + amount_to_transfer_over_msat,
```

Comment:
I'm pretty confused here - previously we included `$next_hops_fee_msat` in the total_fee no matter what, which makes sense - we always pay it (modulo decreasing the value over this path). Now we only include it in `total_fee_msat` if this calculates a fee, which doesn't always happen as `old_entry.src_lowest_inbound_fee` can be a dummy value for some hops.

I think we should add `$next_hops_fee_msat` to `value_left_to_collect_msat`, passing it all the way through.

Reply:
If compute_fee fails, `total_fee_msat = u64::max_value(),`, meaning that we give up on this channel (see the comment).  So there is no point of doing any other manipulations here.

>I think we should add $next_hops_fee_msat to value_left_to_collect_msat, passing it all the way through.

This part I don't understand at all, sorry, could you rephrase?

I also tried to understand your references to other code as well, could you be more precise with old variable names etc?

Maybe a key to understanding your point is understanding this: "as old_entry.src_lowest_inbound_fee can be a dummy value for some hops.". Can you explain this part?

Reply:
I see, you are right. Added it to `amount_to_transfer_over_msat`, just to keep the meaning of `value_left_to_collect_msat`.